spring-cloud-stream-binder-jms-common/src/main/java/org/springframework/cloud/stream/binder/jms/utils/SpecCompliantJmsHeaderMapper.java

```diff
@@ -0,0 +1,42 @@
+package org.springframework.cloud.stream.binder.jms.utils;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import javax.jms.Message;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.integration.jms.DefaultJmsHeaderMapper;
+import org.springframework.messaging.MessageHeaders;
+
+/**
+ * Replace all header names that contain '-' with '_' due to IBM MQ header name
+ * constraints.
+ *
+ * See http://stackoverflow.com/a/30024766/2408961 for context.
+ *
+ * @author Donovan Muller
+ */
+public class SpecCompliantJmsHeaderMapper extends DefaultJmsHeaderMapper {
```

Comment:
We should consider adding this to Spring Integration. In fact, with the upcoming (major) 5.0 release we could make this behavior default (but provide an option to revert to previous behavior in case some brokers are more lenient).